### DISCUSSION TITLE
An error occurred while sending the request.

### DISCUSSION THREAD

**Discussion by grounzero** [dotnet/yarp#710] [Q&A] [answered]
If I refresh a page particularly if it is still loading I get a 502 error.  Is there any way I can capture exceptions and prevent the 502 from being seen in the browser?  The errors can be any of these:

```
System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: The request was aborted.
 ---> System.IO.IOException: The response ended prematurely while waiting for the next frame from the server.
   at System.Net.Http.Http2Connection.<ReadFrameAsync>g__ThrowMissingFrame|48_1()

```

```
System.Threading.Tasks.TaskCanceledException: The operation was canceled.
 ---> System.IO.IOException: Unable to read data from the transport connection: The I/O operation has been aborted because of either a thread exit or an application request..
 ---> System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.
```

```
Request: An error was encountered before receiving a response.
System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.InvalidOperationException: Stream was already consumed.
   at Microsoft.ReverseProxy.Service.Proxy.StreamCopyHttpContent.SerializeToStreamAsync(Stream stream, TransportContext context)
```

**Comment by Tratcher** [comment, marked as answer]
Refreshing a page that's still loading results in the current requests being cancelled and possibly the connections being aborted (HTTP/1.x). We listen to these cancelation events and cancel the proxy requests as well which triggers the error handling logic and the 502s. However, you shouldn't see the 502s in that case sense the client canceled and created new requests. Future requests should not be affected.

That last InvalidOperationException "Stream was already consumed." is surprising. That implies a race condition in the abort logic.

Needs investigation.

**Reply by Tratcher** [reply]
Can you please convert this to an issue?

**Reply by samsp-msft** [reply]
#711

**Reply by grounzero** [reply]
If I refresh the app without going through YARP then I don't see a 502 in the browser.  Is there any way I can avoid this and get the same behaviour going via YARP?

My config is here https://github.com/microsoft/reverse-proxy/issues/711#issuecomment-773710935
